### DISCUSSION TITLE
Expected Performance

### DISCUSSION THREAD

**Discussion by derWebdesigner** [libvips/php-vips#164] [Q&A]
Dear Community,

I just started to implement my own Image Resizer with libvips and after quite a lot of learning steps I finally have everything working in a Docker image and pushed it to DigitalOcean App Plattform. When I have a page with 10 jpg images that are around 2048x1365px and the Image Resizer wants to size them down to 1000px wide webp images, the answer for all 10 images suddenly took between 6-8s when using a single core instance with 1GB RAM. The CPU doesn't even touch the 30% usage mark, so I was wondering if this is what I can expect and if this is realistic. I worked for quite a few days now on this because I thought there is something I did wrong, but maybe I am just having the wrong expectations after it took only 300ms locally for many images but that's on a quite fast MacBookPro M1 Max.

Scaling up from 1 instance to 2 instances cut the processing time in half so that all 10 images now were calculated in 3.5s and therefore I would love to know if you think that simply more small instances are the best choice. My gut tells me so after seeing that the CPU isn't even used a lot and simply makes all parallel requests slower.

Thank you so much for sharing your experience and thank you as well to everybody for working on this awesome project.
Manuel

**Comment by jcupitt** [comment]
Hi @derWebdesigner,

A sudden huge drop in performance usually means you're starting to swap. If you're doing 10 resizes at once I can see it maybe passing your 1gb ram limit.

I would benchmark ram use on your target hardware, then have something to limit the number of resizes you try to run at once to make sure you stay within your memory budget. You might need a separate resize worker process that picks up at most N jobs at once from a DB table.

I tried on this PC (a threadripper pro with a fast NVME drive):

```
$ mkdir test
$ cd test
$ for i in {1..100}; do cp ../k2.jpg $i.jpg; done
$ vipsheader 1.jpg 
1.jpg: 1450x2048 uchar, 3 bands, srgb, jpegload
```

So 100 1450 x 2048 pixel images (close to your case). I see:

```
$ VIPS_CONCURRENCY=1 /usr/bin/time -f %M:%e vipsthumbnail -s 1000 1.jpg
41544:0.05
```

Threading doesn't help a lot with thumbnailing (it's limited by jpeg decode and encode), so you can save a bit of memory by turning down the threadpool. It needs 40mb of memory and c. 50ms. 

With gnu parallel I see:

```
$ rm tn*
$ VIPS_CONCURRENCY=1 time parallel vipsthumbnail -s 1000 {} ::: *.jpg
7.23user 1.72system 0:00.44elapsed 2008%CPU (0avgtext+0avgdata 41584maxresident)k
0inputs+28112outputs (323major+501800minor)pagefaults 0swaps
```

So 400ms to resize 100 images.

Though I don't know anything about digitalocean :( Maybe they have something to throttle processes which use more than a few ms of CPU?

**Reply by derWebdesigner** [reply]
Wow, thank you so much for such detailed feedback, really appreciate it.

I can't believe that it's only 400ms for 100 images. I tracked the exact performance on my local Macbook and even if I only calculate the pure resize (without reading a file because I am loading them from S3) it's still 100-150ms for each photo on an M1 Max with 32GB RAM.

Is the _memory_limit = 1024M_ set in the php.ini an issue for this case or how can I make sure that the RAM is not the limiting factor? I had a look at the other Docker examples here and I really don't see any potential there anymore for making big changes that could be the big blocker. I also upscaled the instance from 1GB RAM to 2GB and it didn't change anything so I guess it's either something wrong about my setup or it's really a limitation by digitalocean although it's also not mega fast locally I think (it would be okay though if the initial generation per image uses 500ms and then the files are cached anyway, but all 7s+ is simply too slow).

Is there any big issue with my code maybe that I am writing them to buffer or that I need to set another flag when using the writeToBuffer function?

```
// read the image from S3
$originalImageAsString = $this->imageStorage->read($image->getOriginalName());
$originalImage         = Vips\Image::newFromBuffer($originalImageAsString);

// resize and return the result immediately to safe the actual file onKernelTerminate to not block the shown image even longer
$vipsImage = $originalImage->thumbnail_image($image->generatedWidth, ['height' => $image->generatedHeight]);
$image->content = $vipsImage->writeToBuffer('.' . $image->getFormat(), ['Q' => $image->getQuality(), 'strip' => true]);
```

**Reply by derWebdesigner** [reply]
Or do I maybe need to put attention to these settings I get when using phpinfo and searching for VIPS? I couldn't find anything related to having to update these settings, but it feels weird to read a 0 for memory currently allocated or is 0 simply unlimited and I need to even reduce the memory for PHP and increase it here somehow? (sorry if this is maybe a dumb question but I am trying for 3 weeks now to get this running and I am super thankful for any kind of hint)

Vips headers version | 8.14.0
-- | --
Vips library version | 8.14.0
Vips ABI version | 58.0.16
Major version | 8
Minor version | 14
Micro version | 0
Cache max mem | 104857600
Cache max operations | 100
Cache current operations | 0
Cache max open files | 100
Memory allocations | 0
Memory currently allocated | 0
Memory high water | 0
Concurrency | 8

**Reply by jcupitt** [reply]
```php
$originalImage         = Vips\Image::newFromBuffer($originalImageAsString);
$vipsImage = $originalImage->thumbnail_image($image->generatedWidth, ['height' => $image->generatedHeight]);
```

Don't use `thumbnail_image`! It's really slow, it's just for emergencies. You'll usually see much better performance with `thumbnail_buffer`.

Set `VIPS_CONCURRENCY` as an environment variable in your startup.

jpeg encode and decode is fast on x86 because it's usually on a vector path. The M1 is still doing it all in software, I think, though I'm sure someone will make a vector path for the M1 soon. I assume digitalocean is x86.

You can download from S3, resize, and upload again all in one operation, have a look at VipsSource and VipsTarget. There are some notes about it here:

https://www.libvips.org/2019/12/11/What's-new-in-8.9.html

**Reply by derWebdesigner** [reply]
@jcupitt That sounds great, I will immediately test that, thank you so much. I will also look at VipsSource. Is there a way to immediately return the content of the image as well before writing it so that I would have to read it, write it and then read it again to return it? I found it really hard to find all the options and available settings via the documentation, so I am relying quite often on what others did post as their solution.

VIPS_CONCURRENCY with any value for the number of threads or simply empty?

**Reply by jcupitt** [reply]
Sure, you can make a VipsTarget which will write directly to the user's browser, if that's what you mean.

I'd set `VIPS_CONCURRENCY` to 1. Or there's php-vips API for this as well, have a look at `Config`.

The php docs are pretty bad, unfortunately. php-doc is a bit limited and won't let us describe all the options. I'd use the main libvips API docs, it's easy to see what the php equivalent would be, eg.:

https://www.libvips.org/API/current/libvips-resample.html#vips-thumbnail

**Reply by derWebdesigner** [reply]
@jcupitt That helped me so much already, thank you dear John. It's a little bit faster, but maybe this is something that needs to be solved via scaling with many server instances. Not sure, but I really don't want to steal more of your time and I will look for somebody who could have a look at it. I guess you are not available for any paid freelance work, right?

**Reply by jcupitt** [reply]
I do some consulting, but I'm fully booked until next June, unfortunately. Plus web server config isn't at all my strong suit, you'd be much better off with someone who specializes in that.

**Reply by derWebdesigner** [reply]
Crazy, but I am super happy for you and fully understand why somebody like you is asked for so much. Thank you so much again, you are awesome and your help in this community for so many years is just amazing.
